mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,691 @@
+        query_entity = QueryEntity.from_query(
+            query=processed_query.query, span=span, entity=entity
+        )
+        return query_entity
+
+    def _resolve_conflicts(self, current_entities, annotated_entities):
+        """ Resolve overlaps between existing entities and newly annotad entities.
+
+        Args:
+            current_entities (list): List of existing query entities.
+            annotated_entities (list): List of new query entities.
+
+        Returns:
+            final_entities (list): List of resolved query entities.
+        """
+        overwrite = self.config["overwrite"]
+        base_entities = annotated_entities if overwrite else current_entities
+        other_entities = current_entities if overwrite else annotated_entities
+
+        additional_entities = []
+        for o_entity in other_entities:
+            no_overlaps = [
+                _no_overlap(o_entity, b_entity) for b_entity in base_entities
+            ]
+            if all(no_overlaps):
+                additional_entities.append(o_entity)
+        base_entities.extend(additional_entities)
+        return base_entities
+
+    # pylint: disable=R0201
+    def _unannotate_query(self, processed_query, remove_entities):
+        """ Removes specified entities in a processed query. If all entities are being
+        removed, this function will not remove entities that the annotator does not support
+        unless it is explicitly specified to do so in the config with the param
+        "unannotate_supported_entities_only" (boolean).
+
+        Args:
+            processed_query (ProcessedQuery): A processed query.
+            remove_entities (list): List of entities to remove.
+        """
+        remove_supported_only = self.config["unannotate_supported_entities_only"]
+        keep_entities = []
+        for query_entity in processed_query.entities:
+            if remove_entities == ["*"]:
+                is_supported_entity = self.valid_entity_check(query_entity.entity.type)
+                if remove_supported_only and not is_supported_entity:
+                    keep_entities.append(query_entity)
+            elif query_entity.entity.type not in remove_entities:
+                keep_entities.append(query_entity)
+        processed_query.entities = tuple(keep_entities)
+
+    @abstractmethod
+    def parse(self, sentence, **kwargs):
+        """ Extract entities from a sentence. Detected entities should be
+        represented as dictionaries with the following keys: "body", "start"
+        (start index), "end" (end index), "value", "dim" (entity type).
+
+        Args:
+            sentence (str): Sentence to detect entities.
+
+        Returns:
+            entities (list): List of entity dictionaries.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class SpacyAnnotator(Annotator):
+    """ (English) Annotator class that uses spacy to generate annotations.
```

Comment:
Let's add the type of the entities that the SpacyAnnotator supports.

Reply:
Added

Reply:
Added